Bind behavior of simple name under 'is' constant pattern

I'm not sure this is by design or not, but simple name bind in constant pattern under `is` expression differs from ordinary member access bind or bind of constant pattern under `case` label:
```c#
using System;

class Foo {
  public ClassVsProp ClassVsProp { get; }

  public void M(object o) {
    if ((string) o == ClassVsProp.Constant) { } // OK
  
    // error CS0176: Member 'ClassVsProp.Constant' cannot be accessed
    // with an instance reference; qualify it with a type name instead
    if (o is ClassVsProp.Constant) { }

    switch (o) {
      case ClassVsProp.Constant: // OK
        break;
    }
  }
}

class ClassVsProp {
  public const string Constant = "abc";
}
```

I think the difference arises because syntax tree shape: simple syntax name node in `o is ClassVsProp.Constant` is not wrapped into constant pattern syntax node, since it's impossible to disambiguate it with ordinary type usage of `is` expressions.

cc @gafter

--- a/src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
@@ -984,7 +984,7 @@ namespace Microsoft.CodeAnalysis.CSharp
         /// </summary>
         private BoundExpression BindQualifiedName(QualifiedNameSyntax node, DiagnosticBag diagnostics)
         {
-            return BindMemberAccessWithBoundLeft(node, this.BindExpression(node.Left, diagnostics), node.Right, node.DotToken, invoked: false, indexed: false, diagnostics: diagnostics);
+            return BindMemberAccessWithBoundLeft(node, this.BindLeftOfPotentialColorColorMemberAccess(node.Left, diagnostics), node.Right, node.DotToken, invoked: false, indexed: false, diagnostics: diagnostics);
         }
 
         private BoundExpression BindParenthesizedExpression(ExpressionSyntax innerExpression, DiagnosticBag diagnostics)

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs
@@ -4324,20 +4324,29 @@ public class Program
         {
             var source =
 @"
-class Foo {
-  public Color Color { get; }
+public class Program
+{
+    public static Color Color { get; }
 
-  public void M(object o) {
-    if (o is Color.Constant) { }
-  }
+    public static void M(object o)
+    {
+        System.Console.WriteLine(o is Color.Constant);
+    }
+
+    public static void Main()
+    {
+        M(Color.Constant);
+    }
 }
 
-class Color {
-  public const string Constant = ""abc"";
+public class Color
+{
+    public const string Constant = ""abc"";
 }
 ";
-            var compilation = CreateCompilationWithMscorlib45(source);
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe);
             compilation.VerifyDiagnostics();
+            var comp = CompileAndVerify(compilation, expectedOutput: "True");
         }
     }
 }

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs
@@ -4318,5 +4318,26 @@ public class Program
             compilation.VerifyDiagnostics();
             var comp = CompileAndVerify(compilation, expectedOutput: "True");
         }
+
+        [Fact]
+        public void ColorColorConstantPattern()
+        {
+            var source =
+@"
+class Foo {
+  public Color Color { get; }
+
+  public void M(object o) {
+    if (o is Color.Constant) { }
+  }
+}
+
+class Color {
+  public const string Constant = ""abc"";
+}
+";
+            var compilation = CreateCompilationWithMscorlib45(source);
+            compilation.VerifyDiagnostics();
+        }
     }
 }

